### ISSUE TITLE
Manage backward compatibility of legacy links

### ISSUE THREAD

**Issue by jolelievre** [PrestaShop/PrestaShop#10745]
**Describe the bug**
Some former controller urls are not correctly managed with le Link class when they have been migrated:

e.g.: 
`getAdminLink('AdminModules')` will work (only if the controller has been added in an ugly array)
`getAdminLink('AdminModules', ['action' => 'configure'])` won't
`getAdminLink('AdminModules', ['configure' => true])` won't
`getAdminLink('AdminModules').'?configure'` won't

This results in lots of legacy urls that will end up as 404 in many modules, but even core pages if they have not been migrated yet

**Solution**
The Link class recently added a feature which is able to link a former controller to a new route through the `_legacy_link` route property:

```
admin_link_block_list:
  path: /link-widget/list
  methods: [GET]
  defaults:
    _controller: 'PrestaShop\Module\LinkList\Controller\Admin\Improve\Design\LinkBlockController::listAction'
    _legacy_controller: AdminLinkWidget
    _legacy_link: AdminLinkWidget
```
It is limited to a single controller for now, so we need to extend this feature to be able to match a `{controller}:{action}` format which would allow this:

```
admin_link_block_create:
  path: /link-widget/create
  methods: [GET]
  defaults:
    _controller: 'PrestaShop\Module\LinkList\Controller\Admin\Improve\Design\LinkBlockController::createAction'
    _legacy_controller: AdminLinkWidget
    _legacy_link: AdminLinkWidget:add
```

Now modifying the Link class to manage the correct transformation of legacy urls to new Symfony urls is a first step, but not enough. Because in many modules developer don't use the `getAdminLink` $params argument they simply concatenate their arguments to the url so they are reachable through $_GET

So to manage the correct redirection of these urls we also need to extend the Dispatcher class which should be able to correctly redirect to the appropriate Symfony route. But in the admin context, symfony routing management occurs before the Dispatcher is called, so an even better solution (which does not remove the Dispatcher one for safety purpose) coud be to add a LegacyRouteListener which is able to correctly redirect before the Dispatcher is called.

To ease the use of this behaviour in these different classes, and for easier testing, of course a dedicated service will have to be implemented.

**In a nutshell**

1. Create a `LegacyRouteMatcher` service able to interpret an url, or an array of parameters to match them with a new symfony routes via the `_legacy_link` route parameter
2. Test this service **thoroughly**
3. Integrate the service in the Link class, and test it
4. Integrate the service in the Dispatcher and test it
5. Create a `LegacyRouteListener` which is able to redirect the legacy routes beforehand, and test it in Integration tests
6. Update every core routing files to add the appropriate `_legacy_link` property
7. Clean the `Link` to remove the associative route matching array as it won't be necessary any more
8. Write the documentation

**Additionnal information**
PrestaShop version: 1.7.5
PHP version: N/A
